### ISSUE TITLE
RangeError: Maximum call stack size exceeded when compileDebug is false

### ISSUE THREAD

**Issue by ghominejad** [jadejs/jade#1846]
```JAVASCRIPT
var jade = require('jade');

var jTemplate = '';

for(var i = 0; i < 2000; i++) {
	jTemplate += 'string!=name\n';
}

var fn = jade.compile(jTemplate, { compileDebug : false});
var html = fn({ name:'my name is ...'});
```

error : 
/data1/Projects/tests/jade-test/node_modules/jade/node_modules/with/node_modules/acorn-globals/node_modules/acorn/util/walk.js:40
    function c(node, st, override) {
              ^
RangeError: Maximum call stack size exceeded

**Comment by ForbesLindesay** [comment]
This is going to be very tricky to fix.  Internally, we use acorn to parse the initial JavaScript we generate in order to figure out the list of all global variables (e.g. `name` in your example).  We use acorn's `util/walk` utility to walk the resulting AST as part of that process.  Unfortunately, but not unsurprisingly, acorn performs a recursive walk of this tree, so trees beyond a certain depth will result in a stack overflow.  The resulting JavaScript from your example template will have something that looks a bit like `a + a + a + a + a + a` (where `a` is the expression that results from `string!=name` and this is repeated 2000 times but you get the idea).  Unfortunately, the AST for this is not an array of items to be added together, but rather a recursive structure like:

```js
{
    "type": "Program",
    "body": [
        {
            "type": "ExpressionStatement",
            "expression": {
                "type": "BinaryExpression",
                "operator": "+",
                "left": {
                    "type": "BinaryExpression",
                    "operator": "+",
                    "left": {
                        "type": "BinaryExpression",
                        "operator": "+",
                        "left": {
                            "type": "BinaryExpression",
                            "operator": "+",
                            "left": {
                                "type": "BinaryExpression",
                                "operator": "+",
                                "left": {
                                    "type": "Identifier",
                                    "name": "a"
                                },
                                "right": {
                                    "type": "Identifier",
                                    "name": "a"
                                }
                            },
                            "right": {
                                "type": "Identifier",
                                "name": "a"
                            }
                        },
                        "right": {
                            "type": "Identifier",
                            "name": "a"
                        }
                    },
                    "right": {
                        "type": "Identifier",
                        "name": "a"
                    }
                },
                "right": {
                    "type": "Identifier",
                    "name": "a"
                }
            }
        }
    ]
}
```

Options for fixing this:

1. Make a version of acorn's walk util that is not recursive.
2. Avoid code that would produce this recursive structure - this could be achieved by pushing the chunks into the array separately, although that does result in marginally less efficient code.  (i.e. generate `buf.push(a);buf.push(a);buf.push(a)` rather than `buf.push(a + a + a)`).

I would prefer option 1, but it's most likely a lot harder to do.  We could consider something like switching from option 1 to option 2 after something like 100 pushes.

**Comment by TimothyGu** [comment]
@ForbesLindesay said:
> 1. Make a version of acorn's walk util that is not recursive.

How? What would the result AST look like?

**Comment by vendethiel** [comment]
The ast doesn't need to change. Tail call rewriting or trampolining can help acorn's issues, tho.

**Comment by TimothyGu** [comment]
@vendethiel Oh OK, I'm not that into CS right now.

**Comment by ForbesLindesay** [comment]
I have written a fix for this in https://github.com/ForbesLindesay/acorn-globals/pull/2

I mostly just wrote that because I found the problem interesting.  There are two problems with this solution that mean it's probably not that useful:

1. The acorn JavaScript parser is recursive as well, so it only actually gives you about another 20% allowance on the stack depth.
2. The V8 JavaScript parser is also recursive, so even if we did somehow make acorn allow longer stacks, we would still get stuck at some point.

And that's without even considering problems like UglifyJS which most people will use in production (another recursive parser, so another limit).

One thing that works, but is a little bit mad is to take advantage of the associativity of addition and rather than writing `a + a + a + a + a + a + a + a` you instead write `((a + a) + (a + a)) + ((a + a) + (a + a))`.  This lets you go way way further (even without modifying acorn/utils/walk it can comfortably parse and walk 2²⁰ (1,048,576) additions.

I think the best bet is to park this till 2.0.0 (as I don't think many people will be hitting this) and then we can add an optimisation to the compiler that forces a break every 100 additions or so.
